### ISSUE TITLE
Sample unit test code doesnt verify diagnostics at line/column start

### ISSUE THREAD

**Issue by pdelvo** [dotnet/roslyn#1641]
The issue is in [here](https://github.com/dotnet/roslyn/blob/b6484300dfafb43af0c27e542ec457a7583e1aa8/src/Samples/CSharp/AsyncPackage/Test/DiagnosticVerifier.cs#L193) and [here](https://github.com/dotnet/roslyn/blob/b6484300dfafb43af0c27e542ec457a7583e1aa8/src/Samples/CSharp/AsyncPackage/Test/DiagnosticVerifier.cs#L204). If the line/column is 0 unit tests will not care if the expected line/column matches that. Also see https://github.com/DotNetAnalyzers/StyleCopAnalyzers/pull/569.

**Comment by sharwell** [comment]
Note that these conditions were intended to allow test code to optionally disable validation for the line and/or column where a diagnostic is reported. The bug is that they check the values for `actualLinePosition` instead of for `expected`. The pull request linked above shows how we corrected this in our local code, but the Roslyn templates should be updated to reflect the same correction.

**Comment by gafter** [comment]
@sharwell It looks like these conditions were intended to avoid testing locations for diagnostics that do not have locations. The comment says

```cs
            // Only check line position if there is an actual line in the real diagnostic
```

What am I missing here? Or are you saying that the comment is just as bad as the code?

**Comment by sharwell** [comment]
@gafter It's probably just as bad. Testing locations for diagnostics that have no location is easy (just set the expected result to the value used to indicate "no location". What's hard (or might need to be handled) is diagnostics that have locations that can vary based on some external state and/or order of operations. Or perhaps cases where the purpose of the test does not rely on location testing.

**Comment by TyOverby** [comment]
@jaredpar: is this still "up for grabs", or should I actually take it?
